src/esp32/board.rs

```diff
@@ -74,6 +78,52 @@ where
             None => Err(anyhow::anyhow!("couldn't find analog reader {}", name)),
         }
     }
+    fn set_power_mode(
+        &self,
+        mode: component::board::v1::PowerMode,
+        duration: Option<Duration>,
+    ) -> anyhow::Result<()> {
+        info!(
+            "Esp32 received request to set power mode to {} for {} milliseconds",
+            mode.as_str_name(),
+            match duration {
+                Some(dur) => dur.as_millis().to_string(),
+                None => "<forever>".to_string(),
+            }
+        );
+
+        anyhow::ensure!(
+            mode == component::board::v1::PowerMode::OfflineDeep,
+            "unimplemented: EspBoard::set_power_mode: modes other than 'OfflineDeep' are not currently supported"
+        );
+
+        if let Some(dur) = duration {
+            let dur_micros = dur.as_micros() as u64;
+            let result: esp_idf_sys::esp_err_t;
+            unsafe {
+                result = esp_idf_sys::esp_sleep_enable_timer_wakeup(dur_micros);
+            }
+            anyhow::ensure!(
+                result == esp_idf_sys::ESP_OK,
+                "unimplemented: EspBoard::set_power_mode: sleep duration {:?} rejected as unsupportedly long", dur
+            );
+            warn!("Esp32 entering deep sleep for {} microseconds!", dur_micros);
+        } else {
+            warn!("Esp32 entering deep sleep without scheduled wakeup!");
+        }
+
+        unsafe {
+            esp_idf_sys::esp_deep_sleep_start();
```

Comment:
Yes, that's a good point. As written we won't answer the rpc on a path that successfully invokes `esp_deep_sleep_start`. Though, if we move the actual logic to enter the offline state to sometime after returning success, there is some chance that we fail to enter the desired sleep mode but will have already reported success. Maybe also a potential for that reply to never actually be sent if the grpc / networking / physical layer does any buffering or async io because the board may just stop executing instructions? So I'd consider any reply as potentially unreliable, but that may be fine in practice. I can look into refactoring a bit to move the call to after RPC reply. It will require a somewhat more complex interaction between `grpc.rs` and the `board` API so I'll think about what that should look like.

Regarding the wifi / bt controls, I had seen that note too, but I hadn't yet chased down what crate exposed them. I'll go do that and try to add the relevant calls.

Reply:
After a quick chat with steve it seems like `esp_deep_sleep_start` does in fact turn off the power to the radio (albeit no gracefully) and you could confirm significant power saving using an USB amp meter. In that case I don't think we should further change the implementation in this PR. we can revisit it later when we add more power levels support